README
gzdec: a GStreamer gzip decoder plugin
######################################

This GStreamer plugin,'gzdec', should receive a stream compressed with gzip and should emit an uncompressed stream. It only has one sink pad and one source pad.


REQUIREMENTS
######################################

The following GZip library is needed:

 - libz.so 


BUILD
######################################
The plugin can be installed locally by using "$HOME" as prefix:

  $ meson --prefix="$HOME" build/
  $ ninja -C build/ install

However be advised that the automatic scan of plugins in the user home
directory won't work under gst-build devenv.


LAUNCH:
######################################

The plugin can be tested launching "test.sh" to prepare the data, where <test.gz> and <output_test> can be configured.

	$ GST_PLUGIN_PATH=build/plugins/ gst-launch-1.0 filesrc location=<test.gz> ! gzdec ! filesink location=<output_test>


NEWS
- WIP. Still not working. Zlib infate() function never reaches Z_STREAM_END state.


test.sh
#!/bin/sh

#### Variables to configure #####
Media_Path="media"
Media_name="CarlsDate_OfficialTrailer.mp4"

#################################

gzName="$Media_Path/$Media_name.gz"

gzip -c $Media_Path\/$Media_name > $gzName



plugins/gstgzdec.h
/* GStreamer
 * Copyright (C) 2023 FIXME <hbravo.teresa@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifndef _GST_GZDEC_H_
#define _GST_GZDEC_H_

#include <zlib.h>
#include <assert.h>

G_BEGIN_DECLS

/* Standard macros for defining types for this element.  */
#define GST_TYPE_GZDEC   (gst_gzdec_get_type())
#define GST_GZDEC(obj)   (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_GZDEC,GstGzdec))
#define GST_GZDEC_CLASS(klass)   (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_GZDEC,GstGzdecClass))
#define GST_IS_GZDEC(obj)   (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_GZDEC))
#define GST_IS_GZDEC_CLASS(obj)   (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_GZDEC))

#define CHUNK 128000 //Size of the output buffer when decoding the stream

typedef struct _GstGzdec GstGzdec;
typedef struct _GstGzdecClass GstGzdecClass;

/* Definition of structure storing data for this element. */
struct _GstGzdec
{  
  GstElement element;

  GstPad *sinkpad, *srcpad;

  z_streamp gzipStrm;

  gboolean silent;
};

/* Standard definition defining a class for this element. */
struct _GstGzdecClass
{
  GstElementClass parent_class;
};

/* Standard function returning type information. */
GType gst_gzdec_get_type (void);

G_END_DECLS

#endif

plugins/gstgzdecplugin.c
/*
 * GStreamer
 * Copyright (C) 2023 FIXME <hbravo.teresa@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <gst/gst.h>
#include "gstgzdec.h"

// static gboolean
// plugin_init (GstPlugin * plugin)
// {
//   // gst_element_register (plugin, "gzdec", GST_RANK_NONE, GST_TYPE_GZDEC);

//   return TRUE;
// }

static gboolean
plugin_init (GstPlugin * plugin)
{
  return GST_ELEMENT_REGISTER (gzdec, plugin);
}

GST_PLUGIN_DEFINE (GST_VERSION_MAJOR, 
                   GST_VERSION_MINOR, gzdec, 
                   "A gzip decoder plugin", 
                   plugin_init, 
                   VERSION, 
                   "Gzdec",  
                   /* FIXME */
                   "GStreamer", 
                   "http://gstreamer.net/")
